File: companies/Amazon/411.minimum-unique-word-abbreviation.go
package leetcode

//
// @lc app=leetcode id=411 lang=golang
//
// [411] Minimum Unique Word Abbreviation
//
// Notes
// Only dictionary words with the target length matter. For each such word,
// build a bitmask of positions where it differs from target. An abbreviation
// mask is valid iff it keeps at least one differing character for every word.
// Enumerate masks, compute abbreviation length, keep the shortest, and build the
// final abbreviation. Time: O(2^m * (m+d)). Space: O(d), m=len(target).
//
// @lc code=start

import "strconv"

func MinAbbreviation411(target string, dictionary []string) string {
	wordLength := len(target)
	relevant := []string{}
	for _, word := range dictionary {
		if len(word) == wordLength {
			relevant = append(relevant, word)
		}
	}
	if len(relevant) == 0 {
		return strconv.Itoa(wordLength)
	}

	differenceMasks := make([]int, 0, len(relevant))
	for _, word := range relevant {
		difference := 0
		for index := 0; index < wordLength; index++ {
			if target[index] != word[index] {
				difference |= 1 << index
			}
		}
		differenceMasks = append(differenceMasks, difference)
	}

	abbreviationLength := func(mask int) int {
		length := 0
		index := 0
		for index < wordLength {
			length++
			if mask&(1<<index) != 0 {
				index++
			} else {
				for index < wordLength && mask&(1<<index) == 0 {
					index++
				}
			}
		}
		return length
	}

	buildAbbreviation := func(mask int) string {
		parts := []byte{}
		abbreviatedCount := 0
		for index := 0; index < wordLength; index++ {
			if mask&(1<<index) != 0 {
				if abbreviatedCount > 0 {
					parts = append(parts, strconv.Itoa(abbreviatedCount)...)
					abbreviatedCount = 0
				}
				parts = append(parts, target[index])
			} else {
				abbreviatedCount++
			}
		}
		if abbreviatedCount > 0 {
			parts = append(parts, strconv.Itoa(abbreviatedCount)...)
		}
		return string(parts)
	}

	bestMask := (1 << wordLength) - 1
	bestLength := wordLength
	for mask := 0; mask < 1<<wordLength; mask++ {
		currentLength := abbreviationLength(mask)
		if currentLength >= bestLength {
			continue
		}
		valid := true
		for _, difference := range differenceMasks {
			if mask&difference == 0 {
				valid = false
				break
			}
		}
		if valid {
			bestMask = mask
			bestLength = currentLength
		}
	}

	return buildAbbreviation(bestMask)
}

// @lc code=end

File: pt-1/411.minimum-unique-word-abbreviation.java
/*
 * @lc app=leetcode id=411 lang=java
 *
 * [411] Minimum Unique Word Abbreviation
 *
 * A kept-character mask defines an abbreviation. For every dictionary word of
 * the same length, build a difference mask; an abbreviation is unique iff its
 * kept mask intersects every difference mask. Enumerate masks and keep the
 * shortest abbreviation.
 *
 * Java note: bit masks fit the LeetCode target length constraint and make
 * conflict checks O(dictionary size).
 *
 * Time: O(2^L * (D + L)). Space: O(D).
 */

import java.util.ArrayList;
import java.util.List;

// @lc code=start
class Solution {
    public String minAbbreviation(String target, String[] dictionary) {
        int length = target.length();
        List<Integer> differenceMasks = new ArrayList<>();

        for (String word : dictionary) {
            if (word.length() != length) {
                continue;
            }
            int diff = 0;
            for (int i = 0; i < length; i++) {
                if (target.charAt(i) != word.charAt(i)) {
                    diff |= 1 << i;
                }
            }
            differenceMasks.add(diff);
        }

        if (differenceMasks.isEmpty()) {
            return Integer.toString(length);
        }

        int bestMask = (1 << length) - 1;
        int bestLength = length;
        for (int mask = 0; mask < (1 << length); mask++) {
            int currentLength = abbreviationLength(mask, length);
            if (currentLength >= bestLength) {
                continue;
            }
            boolean unique = true;
            for (int diff : differenceMasks) {
                if ((mask & diff) == 0) {
                    unique = false;
                    break;
                }
            }
            if (unique) {
                bestMask = mask;
                bestLength = currentLength;
            }
        }
        return buildAbbreviation(target, bestMask);
    }

    private int abbreviationLength(int mask, int wordLength) {
        int length = 0;
        int index = 0;
        while (index < wordLength) {
            length++;
            if ((mask & (1 << index)) != 0) {
                index++;
            } else {
                while (index < wordLength && (mask & (1 << index)) == 0) {
                    index++;
                }
            }
        }
        return length;
    }

    private String buildAbbreviation(String target, int mask) {
        StringBuilder out = new StringBuilder();
        int abbreviated = 0;
        for (int i = 0; i < target.length(); i++) {
            if ((mask & (1 << i)) != 0) {
                if (abbreviated > 0) {
                    out.append(abbreviated);
                    abbreviated = 0;
                }
                out.append(target.charAt(i));
            } else {
                abbreviated++;
            }
        }
        if (abbreviated > 0) {
            out.append(abbreviated);
        }
        return out.toString();
    }
}
// @lc code=end
